clip_benchmark/metrics/linear_probe.py

```diff
@@ -0,0 +1,242 @@
+            assign_learning_rate(param_group, lr)
+    return _lr_adjuster
+
+
+class Featurizer(torch.nn.Module):
+    def __init__(self, model):
+        super().__init__()
+        self.model = model
+
+    def forward(self, input):
+        # note: not sure if we want to train on l2-normalized features
+        image_features = self.model.encode_image(input)
+        image_features = F.normalize(image_features, dim=-1)
+        return image_features
+
+class FeatureDataset(Dataset):
+    def __init__(self, features, targets):
+        self.features = features
+        self.targets = targets
+
+    def __len__(self):
+        return len(self.features)
+
+    def __getitem__(self, i):
+        return self.features[i], self.targets[i]
+
+
+
+def evaluate(model, train_dataloader, dataloader, fewshot_k, batch_size, num_workers, lr, epochs, 
+             model_id, seed, feature_root, device, amp=True, verbose=False):
+    # warning: we currently only support non-multi-label classification datasets.
+    assert device == 'cuda' # need to use cuda for this else too slow
+    # first we need to featurize the dataset, and store the result in feature_root
+    if not os.path.exists(feature_root):
+        os.mkdir(feature_root)
+    feature_dir = os.path.join(feature_root, model_id)
+    if not os.path.exists(feature_dir):
+        os.mkdir(feature_dir)
+
+    
+    featurizer = Featurizer(model).cuda()
+    autocast = torch.cuda.amp.autocast if amp else suppress
+    if not os.path.exists(os.path.join(feature_dir, 'targets_train.pt')):
+        # now we have to cache the features
+        devices = [x for x in range(torch.cuda.device_count())]
+        featurizer = torch.nn.DataParallel(featurizer, device_ids=devices)
+
+        for j, loader in enumerate([dataloader, train_dataloader]):
+            save_str = '_train' if j == 1 else '_val'
+            features = []
+            targets = []
+            num_batches_tracked = 0
+            num_cached = 0
+            with torch.no_grad():
+                for images, target in tqdm(loader):
+                    images = images.to(device)
+
+                    with autocast():
+                        feature = featurizer(images)
+                    
+                    features.append(feature.cpu())
+                    targets.append(target)
+
+                    num_batches_tracked += 1
+                    if (num_batches_tracked % 100) == 0:
+                        features = torch.cat(features)
+                        targets = torch.cat(targets)
+                        
+                        torch.save(features, os.path.join(feature_dir, f'features{save_str}_cache_{num_cached}.pt'))
+                        torch.save(targets, os.path.join(feature_dir, f'targets{save_str}_cache_{num_cached}.pt'))
+                        num_cached += 1
+                        features = []
+                        targets = []
+                        
+            features = torch.cat(features)
```

Comment:
Just one minor thing I noticed when running it, if the number of batches is proportional to 100, `features` and `targets` will be  empty at this point, and `torch.cat(features)`/`torch.cat(targets)` fail with `NotImplementedError: There were no tensor arguments to this function (e.g., you passed an empty list of Tensors)`. The fix would be to just add the latest features to the cache, only if `len(features) > 0`